New event processing changes behaviour of tests which delete objects

In #79, you switched pytest-qt's `waitSignal` to use busy loops because using `QEventLoop` caused a weird segfault.

I recently wrote a test which fails with the new way of looping without realizing it.

The test is making sure that some function behaves properly if an object is already deleted:

```python
@pytest.mark.parametrize('delete', [True, False])
def test_set_register_stylesheet(delete, qtbot, config_stub, caplog):
    config_stub.data = {'fonts': {'foo': 'bar'}, 'colors': {}}
    obj = Obj("{{ font['foo'] }}")

    with caplog.atLevel(9):  # VDEBUG
        style.set_register_stylesheet(obj)

    records = caplog.records()
    assert len(records) == 1
    assert records[0].message == 'stylesheet for Obj: font: bar;'

    assert obj.rendered_stylesheet == 'font: bar;'

    if delete:
        with qtbot.waitSignal(obj.destroyed):
            obj.deleteLater()

    config_stub.data = {'fonts': {'foo': 'baz'}, 'colors': {}}
    style.get_stylesheet.cache_clear()
    config_stub.changed.emit('fonts', 'foo')

    if delete:
        expected = 'font: bar;'
    else:
        expected = 'font: baz;'
    assert obj.rendered_stylesheet == expected
```

The code under test:

```python
def set_register_stylesheet(obj):
    """Set the stylesheet for an object based on it's STYLESHEET attribute.

    Also, register an update when the config is changed.

    Args:
        obj: The object to set the stylesheet for and register.
             Must have a STYLESHEET attribute.
    """
    qss = get_stylesheet(obj.STYLESHEET)
    log.config.vdebug("stylesheet for {}: {}".format(
        obj.__class__.__name__, qss))
    obj.setStyleSheet(qss)
    objreg.get('config').changed.connect(
        functools.partial(update_stylesheet, obj))


def update_stylesheet(obj):
    """Update the stylesheet for obj."""
    if not sip.isdeleted(obj):
        obj.setStyleSheet(get_stylesheet(obj.STYLESHEET))
```

Those tests now fail, presumably because the object never got deleted:


test_set_register_stylesheet[True]:

```
>       assert obj.rendered_stylesheet == expected
E       assert 'font: baz;' == 'font: bar;'
E         - font: baz;
E         ?         ^
E         + font: bar;
E         ?         ^
```

test_set_register_stylesheet[False]:

```
    @pytest.mark.parametrize('delete', [True, False])
    def test_set_register_stylesheet(delete, qtbot, config_stub, caplog):
        config_stub.data = {'fonts': {'foo': 'bar'}, 'colors': {}}
        obj = Obj("{{ font['foo'] }}")
    
        with caplog.atLevel(9):  # VDEBUG
            style.set_register_stylesheet(obj)
    
        records = caplog.records()
        assert len(records) == 1
>       assert records[0].message == 'stylesheet for Obj: font: bar;'
E       assert 'stylesheet f...j: font: baz;' == 'stylesheet fo...j: font: bar;'
E         - stylesheet for Obj: font: baz;
E         ?                             ^
E         + stylesheet for Obj: font: bar;
E         ?                             ^
```

Here I can imagine I'm not the only one with such a kind of test...

I think the best way to implement SignalBlocker would actually be a [QSignalSpy](http://doc.qt.io/qt-5/qsignalspy.html) (which would also make #64 easier) - however it only takes one signal, so `SignalBlocker.connect` and `MultiSignalBlocker` won't work with that...

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,14 +5,6 @@
 - Reduced verbosity when exceptions are captured in virtual methods
   (#77, thanks @The-Compiler).
 
-- Internal refactoring in `waitSignal` and `waitSignals` to use a manual
-  busy loop instead of relying on `QEventLoop` and `QTimer`; depending
-  on the order objects are destroyed in tests the previous implementation
-  could leave dangling signal connections which could cause a crash (noticed
-  this on Windows with PySide and PyQt5). Should not affect tests,
-  but if users notice any behavior change between this version and `1.5.1`
-  please make sure to report it (#80, #79, thanks @The-Compiler for support).
-
 - `QApplication.processEvents()` is now called before and after other fixtures
   and teardown hooks, to better try to avoid non-processed events from leaking 
   from one test to the next. (#67, thanks @The-Compiler). 

--- a/pytestqt/_tests/test_wait_signal.py
+++ b/pytestqt/_tests/test_wait_signal.py
@@ -3,7 +3,7 @@ import time
 
 import pytest
 
-from pytestqt.qt_compat import QtCore, Signal
+from pytestqt.qt_compat import QtCore, Signal, QT_API
 
 
 def test_signal_blocker_exception(qtbot):
@@ -283,3 +283,21 @@ def test_wait_twice(qtbot, timer, multiple, do_timeout, signaller):
             signaller.signal.emit()
         with func(arg):
             signaller.signal.emit()
+
+
+@pytest.mark.skipif(QT_API == 'pyside', reason='test crashes PySide')
+def test_destroyed(qtbot):
+    """Test that waitSignal works with the destroyed signal (#82).
+
+    For some reason, this crashes PySide although it seems perfectly fine code.
+    """
+    import sip
+
+    class Obj(QtCore.QObject):
+        pass
+
+    obj = Obj()
+    with qtbot.waitSignal(obj.destroyed):
+        obj.deleteLater()
+
+    assert sip.isdeleted(obj)

--- a/pytestqt/plugin.py
+++ b/pytestqt/plugin.py
@@ -351,9 +351,16 @@ class _AbstractSignalBlocker(object):
     """
 
     def __init__(self, timeout=1000, raising=False):
+        self._loop = QtCore.QEventLoop()
         self.timeout = timeout
         self.signal_triggered = False
         self.raising = raising
+        if timeout is None:
+            self._timer = None
+        else:
+            self._timer = QtCore.QTimer(self._loop)
+            self._timer.setSingleShot(True)
+            self._timer.setInterval(timeout)
 
     def wait(self):
         """
@@ -362,26 +369,31 @@ class _AbstractSignalBlocker(object):
         :raise ValueError: if no signals are connected and timeout is None; in
             this case it would wait forever.
         """
-        try:
-            if self.signal_triggered:
-                return
-            if self.timeout is None and not self._signals:
-                raise ValueError("No signals or timeout specified.")
-            timer = QtCore.QElapsedTimer()
-            if self.timeout:
-                timer.start()
-            while not self.signal_triggered:
-                QApplication.instance().processEvents()
-                if self.timeout and timer.hasExpired(self.timeout):
-                    break
-            if not self.signal_triggered and self.raising:
-                raise SignalTimeoutError("Didn't get signal after %sms." %
-                                         self.timeout)
-        finally:
-            self._cleanup()
+        if self.signal_triggered:
+            return
+        if self.timeout is None and not self._signals:
+            raise ValueError("No signals or timeout specified.")
+        if self._timer is not None:
+            self._timer.timeout.connect(self._quit_loop_by_timeout)
+            self._timer.start()
+        self._loop.exec_()
+        if not self.signal_triggered and self.raising:
+            raise SignalTimeoutError("Didn't get signal after %sms." %
+                                     self.timeout)
+
+    def _quit_loop_by_timeout(self):
+        self._loop.quit()
+        self._cleanup()
 
     def _cleanup(self):
-        """clean instance variables and disconnect all signals."""
+        if self._timer is not None:
+            try:
+                self._timer.timeout.disconnect(self._quit_loop_by_timeout)
+            except (TypeError, RuntimeError):
+                # already disconnected by Qt?
+                pass
+            self._timer.stop()
+            self._timer = None
 
     def __enter__(self):
         return self
@@ -434,6 +446,8 @@ class SignalBlocker(_AbstractSignalBlocker):
         quits the event loop and marks that we finished because of a signal.
         """
         self.signal_triggered = True
+        self._loop.quit()
+        self._cleanup()
 
     def _cleanup(self):
         super(SignalBlocker, self)._cleanup()
@@ -443,7 +457,6 @@ class SignalBlocker(_AbstractSignalBlocker):
             except (TypeError, RuntimeError):
                 # already disconnected by Qt?
                 pass
-        self._signals[:] = []
 
 
 class MultiSignalBlocker(_AbstractSignalBlocker):
@@ -463,7 +476,6 @@ class MultiSignalBlocker(_AbstractSignalBlocker):
     def __init__(self, timeout=1000, raising=False):
         super(MultiSignalBlocker, self).__init__(timeout, raising=raising)
         self._signals = {}
-        self._slots = {}
 
     def _add_signal(self, signal):
         """
@@ -473,9 +485,7 @@ class MultiSignalBlocker(_AbstractSignalBlocker):
         :param signal: QtCore.Signal
         """
         self._signals[signal] = False
-        slot = functools.partial(self._signal_emitted, signal)
-        self._slots[signal] = slot
-        signal.connect(slot)
+        signal.connect(functools.partial(self._signal_emitted, signal))
 
     def _signal_emitted(self, signal):
         """
@@ -487,17 +497,7 @@ class MultiSignalBlocker(_AbstractSignalBlocker):
         self._signals[signal] = True
         if all(self._signals.values()):
             self.signal_triggered = True
-
-    def _cleanup(self):
-        super(MultiSignalBlocker, self)._cleanup()
-        for signal, slot in self._slots.items():
-            try:
-                signal.disconnect(slot)
-            except (TypeError, RuntimeError):
-                # already disconnected by Qt?
-                pass
-        self._slots.clear()
-        self._signals.clear()
+            self._loop.quit()
 
 
 class SignalTimeoutError(Exception):

